### DISCUSSION TITLE
Generic Interface Type with ILogger Interface is own class

### DISCUSSION THREAD

**Discussion by ShervanN** [ProfessionalCSharp/ProfessionalCSharp2021#187] [Q&A] [answered]
Hello,

I have seen many times in your book that in the constructor you use your own class. For example:
On 534 ❘ CHAPTER 19 Networking,  

https://ibb.co/txBG8mb

Could you please explain what is used for? when and why should I use it?

```
    public class QuotesServer
    {
        private readonly int _port;
        private readonly ILogger _logger;
        private readonly string _quotesPath;
        private string[]? _quotes;
        private Random _random = new();

        public QuotesServer(IOptions<QuotesServerOptions> options, ILogger<QuotesServer> logger)
        {

        }
    }
}
```

**Comment by christiannagel** [comment, marked as answer]
Using the `ILogger` interface, you cannot inject the non-generic version of this interface. You'll get an error with this. The `ILogger` interface needs a name to use this for logging - this way you can find where log messages relate to. With this, you can either inject `ILoggerFactory`, and pass a name using the `CreateLogger` method - or inject the `ILogger` interface with a generic parameter using a class. The `ILogger` implementation than uses the name of the class as a category name for logging - e.g. in the code snippet the name "QuotesServer".
